threedi_model_migration/api_utils.py

```diff
@@ -0,0 +1,225 @@
+
+import json
+import logging
+import time
+import zipfile
+
+
+logger = logging.getLogger(__name__)
+
+
+def get_or_create_schematisation(
+    api: V3BetaApi, schematisation: Schematisation, overwrite: bool = False
+) -> OASchematisation:
+    resp = api.schematisations_list(name=schematisation.name, owner__unique_id=schematisation.metadata.owner)
+    if resp.count == 1 and not overwrite:
+        logger.info(
+            f"Schematisation '{schematisation.name}' already exists, skipping creation."
+        )
+        return resp.results[0], False
+    elif resp.count == 1 and overwrite:
+        logger.info(
+            f"Schematisation '{schematisation.name}' already exists, deleting..."
+        )
+        delete_schematisation(api, resp.results[0].id)
+
+    logger.info(f"Creating schematisation '{schematisation.slug}'...")
+    obj = OASchematisation(
+        owner=schematisation.metadata.owner,
+        name=schematisation.name,
+        tags=[],
+        meta=schematisation.metadata.meta,
+        created_by=schematisation.metadata.created_by,
+        created=schematisation.metadata.created,
+    )
+    for _ in range(2):
+        try:
+            resp = api.schematisations_create(obj)
+        except ApiException as e:
+            if e.status == 400:
+                errors = json.loads(e.body)
+                if len(errors.get("created_by", [])) == 1:
+                    logger.warning(errors["created_by"])
+                    obj.created_by = None
+                    continue  # try again
+            raise e
+        break
+    return resp, True
+
+
+def delete_schematisation(api: V3BetaApi, schema_id: int):
+    # first delete the revisions
+    while True:
+        resp = api.schematisations_revisions_list(schema_id)
+        if len(resp.results) == 0:
+            break
+
+        for oa_revision in resp.results:
+            api.schematisations_revisions_delete(
+                oa_revision.id, schema_id, {"number": oa_revision.number}
+            )
+            if oa_revision.archived is None:
+                api.schematisations_revisions_delete(
+                    oa_revision.id, schema_id, {"number": oa_revision.number}
+                )
+
+    # then the schematisation
+    api.schematisations_delete(schema_id)
+
+
+def _match_revision(
+    oa_revision: OARevision, revisions: List[SchemaRevision]
+) -> SchemaRevision:
+    """Match an external (OpenAPI) revision with an internal (SchemaRevision)
+
+    The revision with the same number is returned. The commit date is
+    also compared.
+
+    Revisions should be sorted new to old ('revision_nr' descending)
+    """
+    for revision in revisions:
+        if oa_revision.number < revision.revision_nr:
+            break
+        elif oa_revision.number > revision.revision_nr:
+            continue
+        if oa_revision.commit_date == revision.last_update:
+            return revision
+        else:
+            break
+
+
+def get_latest_revision(
```

Comment:
Note: `/v3-beta/schematisations/{id}/latest_revision/` also exists.

Reply:
Yes, but that could include a commit done via the new route (while we want the latest commit from the modeldatabank)